src/ucs/vfs/base/vfs_obj.c

```diff
@@ -257,17 +327,138 @@ void ucs_vfs_obj_remove(void *obj)
 
     node = ucs_vfs_node_find_by_obj(obj);
     if (node != NULL) {
-        ucs_vfs_node_remove(node);
+        ucs_vfs_node_decrease_refcount(node);
+    }
+
+    ucs_spin_unlock(&ucs_vfs_obj_context.lock);
+}
+
+void ucs_vfs_obj_set_dirty(void *obj, ucs_vfs_refresh_cb_t refresh_cb)
+{
+    ucs_vfs_node_t *node;
+
+    ucs_spin_lock(&ucs_vfs_obj_context.lock);
+
+    node = ucs_vfs_node_find_by_obj(obj);
+    if (node != NULL) {
+        node->dirty      = 1;
+        node->refresh_cb = refresh_cb;
     }
 
     ucs_spin_unlock(&ucs_vfs_obj_context.lock);
 }
 
+ucs_status_t ucs_vfs_path_get_info(const char *path, ucs_vfs_path_info_t *info)
+{
+    ucs_string_buffer_t strb;
+    ucs_vfs_node_t *node;
+    ucs_status_t status;
+
+    ucs_spin_lock(&ucs_vfs_obj_context.lock);
+
+    node = ucs_vfs_node_find_by_path(path);
+    if ((ucs_vfs_check_node(node, UCS_VFS_NODE_TYPE_DIR) == 0) &&
+        (ucs_vfs_check_node(node, UCS_VFS_NODE_TYPE_SUBDIR) == 0) &&
+        (ucs_vfs_check_node(node, UCS_VFS_NODE_TYPE_RO_FILE) == 0)) {
+        status = UCS_ERR_NO_ELEM;
+        goto out_unlock;
+    }
+
+    ucs_vfs_node_increase_refcount(node);
+
+    if (node->type == UCS_VFS_NODE_TYPE_RO_FILE) {
+        ucs_string_buffer_init(&strb);
+        ucs_vfs_read_ro_file(node, &strb);
+        info->mode = S_IFREG | S_IRUSR;
+        info->size = ucs_string_buffer_length(&strb);
+        ucs_string_buffer_cleanup(&strb);
+    } else {
+        ucs_vfs_refresh_dir(node);
+        info->mode = S_IFDIR | S_IRUSR | S_IXUSR;
+        info->size = ucs_list_length(&node->children);
+    }
+    status = UCS_OK;
+
+    ucs_vfs_node_decrease_refcount(node);
+
+out_unlock:
+    ucs_spin_unlock(&ucs_vfs_obj_context.lock);
+
+    return status;
+}
+
+char *ucs_vfs_path_read_file(const char *path)
+{
+    ucs_string_buffer_t strb;
+    ucs_vfs_node_t *node;
+    char *result;
+
+    ucs_spin_lock(&ucs_vfs_obj_context.lock);
+
+    node = ucs_vfs_node_find_by_path(path);
+    if (ucs_vfs_check_node(node, UCS_VFS_NODE_TYPE_RO_FILE) == 0) {
+        result = NULL;
+        goto out_unlock;
+    }
+
+    ucs_vfs_node_increase_refcount(node);
+
+    ucs_string_buffer_init(&strb);
+    ucs_vfs_read_ro_file(node, &strb);
+    result = ucs_strdup(ucs_string_buffer_cstr(&strb), "vfs_text");
```

Comment:
can we make `*strb` an in/out parameter for ucs_vfs_path_read_file, to avoid extra memory copy?